devicetypes/smartthings/zwave-rgbw-bulb.src/zwave-rgbw-bulb.groovy

```diff
@@ -0,0 +1,324 @@
+	def linkText = device.label ?: device.name
+	[linkText: linkText, descriptionText: "$linkText: $cmd", displayed: false]
+}
+
+def buildOffOnEvent(cmd){
+	[zwave.basicV1.basicSet(value: cmd), zwave.switchMultilevelV3.switchMultilevelGet()]
+}
+
+def on() {
+	commands(buildOffOnEvent(0xFF), 5000)
+}
+
+def off() {
+	commands(buildOffOnEvent(0x00), 5000)
+}
+
+def refresh() {
+	commands([zwave.switchMultilevelV3.switchMultilevelGet()] + queryAllColors())
+}
+
+def ping() {
+	log.debug "ping().."
+	unschedule(offlinePing)
+	runEvery30Minutes(offlinePing)
+	command(zwave.switchMultilevelV3.switchMultilevelGet())
+}
+
+def offlinePing() {
+	log.debug "offlinePing()..."
+	sendHubCommand(new physicalgraph.device.HubAction(command(zwave.switchMultilevelV3.switchMultilevelGet())))
+}
+
+def setLevel(level) {
+	setLevel(level, 1)
+}
+
+def setLevel(level, duration) {
+	log.debug "setLevel($level, $duration)"
+	if(level > 99) level = 99
+	commands([
+		zwave.switchMultilevelV3.switchMultilevelSet(value: level, dimmingDuration: duration),
+		zwave.switchMultilevelV3.switchMultilevelGet(),
+	], 5000)
+}
+
+def setSaturation(percent) {
+	log.debug "setSaturation($percent)"
+	setColor(saturation: percent)
+}
+
+def setHue(value) {
+	log.debug "setHue($value)"
+	setColor(hue: value)
+}
+
+def setColor(value) {
+	log.debug "setColor($value)"
+	def rgb
+	if (state.staged == null) {
+		state.staged = [:]
+	}
+	if (value.hex) {
+		state.staged << [color: value.hex] // stage ST RGB color attribute
+		def hsv = colorUtil.hexToHsv(value.hex) // convert to HSV
+		state.staged << [hue: hsv[0], saturation: hsv[1]] // stage ST hue and saturation attributes
+		rgb = value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) } // separate RGB elements for zwave setter
+	} else {
+		state.staged << value.subMap("hue", "saturation") // stage ST hue and saturation attributes
+		def hex = colorUtil.hsvToHex(Math.round(value.hue) as int, Math.round(value.saturation) as int) // convert to hex
+		state.staged << [color: hex] // statge ST RGB color attribute
+		rgb = colorUtil.hexToRgb(hex) // separate RGB elements for zwave setter
+	}
+	def cmds = [zwave.switchColorV3.switchColorSet(red: rgb[0], green: rgb[1], blue: rgb[2], warmWhite: 0, coldWhite: 0),
+		        zwave.switchColorV3.switchColorGet(colorComponent: RGB_NAMES[0])]
+	commands(cmds)
+}
+
+private tempToZwaveWarmWhite(temp) {
+	temp = temp < COLOR_TEMP_MIN ? COLOR_TEMP_MIN : temp > COLOR_TEMP_MAX ? COLOR_TEMP_MAX : temp
+	def warmValue = ((COLOR_TEMP_MAX - temp) / COLOR_TEMP_DIFF * WHITE_MAX) as Integer
+}
+
+private tempToZwaveColdWhite(temp) {
+	(WHITE_MAX - tempToZwaveWarmWhite(temp))
+}
+
+def setColorTemperature(temp) {
+	log.debug "setColorTemperature($temp)"
+	def warmValue = tempToZwaveWarmWhite(temp)
+	def coldValue = tempToZwaveColdWhite(temp)
+	if (state.staged == null) {
+		state.staged = [:]
+	}
+	state.staged << [colorTemperature: temp] // stage ST colorTemperature attribute
```

Comment:
Shouldn't this be an `=` instead of `<<`? I think we want to overwrite the staged value

Reply:
With the consistency guarantees around read and write of state from the DB and how this interacts with DTH instantiation, I found that it was not safe to have the temp and color handlers clobbering each others' cached state.  This is definitely intentional.

The code attempts to target creation of emitted events by keying these on receipt of zwave RGB color information vs. receipt of zwave coldWhite/warmWhite information.